python/lsst/pipe/tasks/makeWarp.py

```diff
@@ -576,6 +592,7 @@ def prepareCalibratedExposures(self, calExpList, backgroundList=None, skyCorrLis
                 calexp.maskedImage -= skyCorr.getImage()
 
             indices.append(index)
+            calExpList[index] = calexp
```

Comment:
This makes me nervous ... this is an update of the list as you enumerate over it?

But more importantly, I'm not sure why this API has to change at all.  Why can't we still read the calexps in the runQuantum and just filter here?  This might require some discussion.

Reply:
Ok, so I think I know where the confusion lies (and am not fully convinced what I'm doing is ideal, but it seemed the best option at the time).  The main difference is that I'm now passing `calExpList` to `prepareCalibratedExposures()` as a `list` of  `DeferredDatasetHandle`s.  The reason for the delayed read is to potentially avoid unnecessarily reading in `calexps` that will never be used (via the `if externalSkyWcsCatalog is None and wcsList[index] is None:` clause at the beginning of the loop...and this is also how a `calexp` with a failed SFM WCS could still pick up the external calibration, if present).  Also, this way the "within patch" assessment is made on the final WCS (externally calibrated, if relevant), not just the one from SFM (I _think_ this is slightly preferred, although probably won't play a huge role in most circumstances).  Since the `calExpList` `list` was of `DeferredDatasetHandle`s, I update it here to contain the (now re-calibrated) actual `calexp`  `Exposure` object.

Does that make any sense?!